app/src/main/java/net/squanchy/support/widget/InterceptingBottomNavigationView.kt

```diff
@@ -0,0 +1,143 @@
+    private var colorProvider: ColorProvider? = null
+
+    var revealDurationMillis: Int = 0
+
+    private val bottomNavigationMenuView: BottomNavigationMenuView
+        get() = getChildAt(0) as BottomNavigationMenuView
+
+    init {
+        super.setOnNavigationItemSelectedListener { onItemSelected(it) }
+        revealDurationMillis = resources.getInteger(android.R.integer.config_shortAnimTime)
+    }
+
+    private fun onItemSelected(menuItem: MenuItem): Boolean {
+        var itemSelected = true
+        if (listener != null) {
+            itemSelected = listener!!.onNavigationItemSelected(menuItem)
+        }
+
+        if (itemSelected && colorProvider != null) {
+            startCircularReveal(colorProvider!!.selectedItemColor, menuItem)
+        }
+
+        return itemSelected
+    }
+
+    private fun startCircularReveal(@ColorInt color: Int, menuItem: MenuItem) {
+        val revealDrawable = background as CircularRevealDrawable
+        setBackgroundHotspot(revealDrawable, menuItem)
+        revealDrawable.animateToColor(color, revealDurationMillis)
+    }
+
+    private fun setBackgroundHotspot(revealDrawable: CircularRevealDrawable, menuItem: MenuItem) {
+        val (x, y) = lastUpEvent?.let { Hotspot.fromMotionEvent(it) } ?: getHotspotFor(menuItem)
+        revealDrawable.setHotspot(x, y)
+    }
+
+    private fun getHotspotFor(menuItem: MenuItem): Hotspot {
+        val selectedPosition = positionFor(menuItem)
+        val menuView = bottomNavigationMenuView
+        val selectedItemView = menuView.getChildAt(selectedPosition)
+
+        return Hotspot.fromCenterOf(selectedItemView)
+            .offsetToParent(menuView)
+    }
+
+    @IntRange(from = 0)
+    private fun positionFor(menuItem: MenuItem): Int {
+        val menuSize = menu.size()
+        for (i in 0 until menuSize) {
+            val itemId = menu.getItem(i).itemId
+            if (itemId == menuItem.itemId) {
+                return i
+            }
+        }
+        throw IllegalArgumentException("Menu item $menuItem not found in the bottom bar items")
+    }
+
+    override fun setBackgroundColor(@ColorInt color: Int) {
+        val background = background
+        if (background is CircularRevealDrawable) {
+            updateBackgroundColor(background, color)
+        } else {
+            setBackgroundColor(color)
+        }
+    }
+
+    override fun setBackground(background: Drawable) {
+        if (background is ColorDrawable) {
+            val currentBackground = getBackground()
+            val newColor = background.color
+            updateBackgroundColor(currentBackground, newColor)
+        } else {
+            throw IllegalArgumentException("Only ColorDrawables are supported")
+        }
+    }
+
+    private fun updateBackgroundColor(drawable: Drawable, newColor: Int) {
+        if (drawable is CircularRevealDrawable) {
+            drawable.color = newColor
```

Comment:
We probably should put an `invalidate()` here too, I may have forgotten it back then D:

Reply:
👍